### PR TITLE
Temp save

### PR DESCRIPTION
**Pull request by henpe071**
When creating a new project, everything is created in a temp folder. Clicking save will move the project to the assign folder.  Open project will also "open" the project in a temp folder. all changes made to the project are only saved in this temp folder and only recorded in the correct folder upon clicking save. 

When clicking save, a pop-up will appear if a project already exist in the assigned folder asking if you're sure you wanna overwrite the existing project. 

I might also have done some small cleanups in the code concerning some brackets and variable names.

### CODE REVIEW

File: ViAn/GUI/projectwidget.cpp [hunk 1]
```diff
@@ -582,14 +583,26 @@ void ProjectWidget::create_folder_item() {
  */
 void ProjectWidget::save_project() {
     if (m_proj == nullptr ) return;
+    QDir dir(QString::fromStdString(m_proj->get_dir()));
+    if (dir.exists()) {
+        QMessageBox msg_box;
```
Comment (fredrikiselius) [inline ViAn/GUI/projectwidget.cpp]: We've added quite alot of message boxes. We should refactor this type of code snippet into a function,

File: ViAn/Project/project.cpp [hunk 2]
```diff
@@ -4,17 +4,20 @@
...
+
 //TODO Fix all these function names
-Project* Project::fromFile(const std::string &full_path)
-{
+Project* Project::fromFile(const std::string &full_path){
     Project* proj = new Project();
     proj->load_saveable(full_path);
     // ensure changes to paths are saved
     proj->save_saveable(full_path);
+    if(proj->tmp_dir.isValid()){
```
Comment (fredrikiselius) [inline ViAn/Project/project.cpp:16]: And if its not? isValid() returns false if the temporary dir could not be created. This needs to be handled somehow. Wont it try to save to a nonexistent directory if it fails?

File: ViAn/Project/project.cpp [hunk 3]
```diff
@@ -25,10 +28,15 @@ Project* Project::fromFile(const std::string &full_path)
  * @param vid_path := path for videos folder
  */
 Project::Project(const std::string& name, const std::string& dir_path){
-    m_name = name;    
-    m_dir = dir_path + "/" + name + "/";
-    m_dir_bookmarks = m_dir + "Bookmarks/";
-    m_file = m_dir + name + ".vian";
+    if(tmp_dir.isValid()){
```
Comment (fredrikiselius) [inline ViAn/Project/project.cpp]: Same with this isValid()

File: ViAn/GUI/projectwidget.cpp [hunk 4]
```diff
@@ -45,6 +49,7 @@ ProjectWidget::ProjectWidget(QWidget *parent) : QTreeWidget(parent) {
 void ProjectWidget::new_project() {
     ProjectDialog* proj_dialog = new ProjectDialog();
     QObject::connect(proj_dialog, SIGNAL(project_path(QString, QString)), this, SLOT(add_project(QString, QString)));
+    delete proj_dialog;
```
Comment (fredrikiselius) [inline ViAn/GUI/projectwidget.cpp]: The new project dialog is closed directly after it opens due to the delete here.

File: ViAn/GUI/projectwidget.cpp [hunk 5]
```diff
@@ -582,14 +583,26 @@ void ProjectWidget::create_folder_item() {
  */
 void ProjectWidget::save_project() {
     if (m_proj == nullptr ) return;
+    QDir dir(QString::fromStdString(m_proj->get_dir()));
```
Comment (fredrikiselius) [inline ViAn/GUI/projectwidget.cpp]: This popup is triggered on each save. It should only happen when the project has never been saved to that location before.

File: ViAn/GUI/projectwidget.cpp [hunk 6]
```diff
@@ -582,14 +583,26 @@ void ProjectWidget::create_folder_item() {
...
     ProjectTreeState tree_state;
     tree_state.set_tree(invisibleRootItem());
-    tree_state.save_state(m_proj->getDir() + "treestate");
+    tree_state.save_state(m_proj->get_tmp_dir() + "treestate");
     m_proj->save_project();
+    m_proj->move_project_from_tmp();
     RecentProject rp;
     rp.load_recent();
-    rp.update_recent(m_proj->getName(), m_proj->full_path());
+    rp.update_recent(m_proj->get_name(), m_proj->get_file());
```
Comment (fredrikiselius) [inline ViAn/GUI/projectwidget.cpp:611]: This no longer updates the recent projects list

**Review by fredrikiselius — CHANGES_REQUESTED** [review]
The new additions broke both the new project window and recent project window for me. See code specific comments

**Comment by henpe071** [comment]
Now the things should be fixed and I also fixed a small thing I missed before.

File: ViAn/GUI/projectwidget.cpp [hunk 7]
```diff
@@ -60,11 +60,17 @@ void ProjectWidget::new_project() {
...
  */
 void ProjectWidget::add_project(QString project_name, QString project_path) {
-    close_project();
     std::string name = project_name.toStdString();
     std::string path = project_path.toStdString();
-    parentWidget()->parentWidget()->setWindowTitle(project_name);
-    m_proj = new Project(name, path);
+    Project* new_proj = new Project(name, path);
+    if (!new_proj->tmp_dir_valid) {
+        delete m_proj;
```
Comment (fredrikiselius) [inline ViAn/GUI/projectwidget.cpp]: Won't this put the program in a state without a project?

File: ViAn/Project/recentproject.cpp [hunk 8]
```diff
@@ -15,9 +15,10 @@ void RecentProject::update_recent(const std::string& name, const std::string &pr
     std::pair<std::string, std::string> new_proj = std::make_pair(name, project_path);
     auto item = std::find(recent_items.begin(), recent_items.end(), new_proj);
     if (item != recent_items.end()) recent_items.erase(item);
+    if (recent_items.size() >= 10) recent_items.resize(10);
```
Comment (fredrikiselius) [inline ViAn/Project/recentproject.cpp]: 10 should be made a constant

**Review by fredrikiselius — APPROVED** [review]

**Pull request merged**